java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java

```diff
@@ -621,8 +621,14 @@ private K disjunctResults(List<K> results) {
                 if (guarded) {
                     ConstrainedTerm result = applySpecRules(term, specRules);
                     if (result != null) {
-                        if (visited.add(result))
+                        if (visited.add(result)) {
                             nextQueue.add(result);
+                        } else {
+                            if (term.equals(result)) {
+                                System.err.println("Step " + step + ": infinite loop after applying a spec rule.");
```

Comment:
This is not an appropriate way to report a warning to a user, please look into KExceptionManager.registerCriticalWarning.